pieces/parallelism/README.md

```diff
@@ -0,0 +1,238 @@
+> correctly. To fix this, modify the header to include a dummy `_` field at the start of the header
+> as can be seen in the example data provided in this piece.
+
+## Output
+
+The outputs are a set of CSV files with the `_processed` suffix, containing the following fields:
+
+* `id`: Unique identifier for the article
+* `publication`: Name of the publication
+* `author`: Name of the author
+* `date`: Date of publication
+* `num_male_pronouns`: Number of male pronouns in the article's text
+* `num_female_pronouns`: Number of female pronouns in the article's text
+
+We also display the run time comparisons of the parallel processing code in Rust and Python, for this
+sample dataset in this repo, as well as the full dataset downloaded from Kaggle.
+
+## Python Setup
+
+Install the dependencies in a virtual environment via `requirements.txt`.
+
+```bash
+# First time setup
+python -m venv venv
+source venv/bin/activate
+pip install -r requirements.txt
+
+# For subsequent runs, simply activate the environment
+source venv/bin/activate
+```
+
+### Run scripts
+
+A single script `main.py` is used to run the parallel processing code, that divides up each CSV's
+records into batches and processes them via `concurrent.futures.ProcessPoolExecutor`.
+
+The script is run as follows:
+
+```bash
+python main.py
+```
+
+### Run tests
+
+Tests can be run as follows:
+
+```bash
+$ pytest -v
+======================================= test session starts ========================================
+platform darwin -- Python 3.11.7, pytest-8.0.0, pluggy-1.4.0 -- /Users/prrao/.pyenv/versions/3.11.7/bin/python3.11
+cachedir: .pytest_cache
+rootdir: /Users/prrao/code/rustinpieces/pieces/parallelism/python
+plugins: Faker-22.0.0
+collected 5 items
+
+test_main.py::test_record PASSED                                                             [ 20%]
+test_main.py::test_count_gendered_pronouns PASSED                                            [ 40%]
+test_main.py::test_clean_text PASSED                                                         [ 60%]
+test_main.py::test_calculate_counts PASSED                                                   [ 80%]
+test_main.py::test_process_batches PASSED                                                    [100%]
+
+======================================== 5 passed in 0.15s =========================================
+```
+
+### Results
+
+> [!NOTE]
+> The timing numbers shown below are the run times from a 2023 M3 Macbook Pro with 32GB of RAM.
+> The Python version used was `3.11.7`.
+
+
+| numRecords | Python     | Rust      | Speedup factor |
+| ---------- | ---------- | --------- | -------------- |
+| 10         | 0.255 sec  | 0.192 sec | 1.3x           |
+| 143000     | 13.493 sec | 3.416 sec | 4.0x           |
+
+For the full dataset, the Rust code using `rayon` is about 4x faster than the Python code. This
+performance difference would only increase as the number of records increases.
```

Comment:
Instead of having this twice could we move this one and the one from the Rust section to the `Takeaways` so it doesn't repeat?

Reply:
Makes sense. That does seem quite redundant now that I look at it again.